### PR TITLE
SONARCS-832 'ref' and 'out' parameters should not be used.

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                c =>
+                {
+                    var parameter = (ParameterSyntax)c.Node;
+
+                    if (!parameter.Modifiers.Any(IsRefOrOut))
+                    {
+                        return;
+                    }
+
+                    var modifier = parameter.Modifiers.First(IsRefOrOut);
```
Comment (Evangelink) [inline src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs]: I would create some helper and merge this statement with the previous `if`.

File: src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            return method.Name.StartsWith("Try", StringComparison.Ordinal) &&
+                method.ReturnType.Is(KnownType.System_Boolean) && 
+                modifier.IsKind(SyntaxKind.OutKeyword);
+        }
+
+        private IMethodSymbol GetContainingMethod(ParameterSyntax parameter, SemanticModel semanticModel)
+        {
+            var method = parameter.GetParents().OfType<MethodDeclarationSyntax>().FirstOrDefault();
+            if (method != null)
```
Comment (Evangelink) [inline src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs]: Merge all the following using the ternary operator.

File: src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        private bool IsTryPattern(IMethodSymbol method, SyntaxToken modifier)
+        {
+            return method.Name.StartsWith("Try", StringComparison.Ordinal) &&
+                method.ReturnType.Is(KnownType.System_Boolean) && 
+                modifier.IsKind(SyntaxKind.OutKeyword);
+        }
+
+        private IMethodSymbol GetContainingMethod(ParameterSyntax parameter, SemanticModel semanticModel)
+        {
+            var method = parameter.GetParents().OfType<MethodDeclarationSyntax>().FirstOrDefault();
```
Comment (Evangelink) [inline src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs]: Could you make a test with dumb parenthesis around some of the parameters? I am not sure your method is going to handle the case. Also I don't think you need the method to retrieve the declaring method (have a look at LiveVariableAnalysis.cs on line 188), Tamas was retrieving the parent without your helper method.

File: src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs (hunk 1 above)
Comment (valhristov) [inline src/SonarAnalyzer.CSharp/Rules/DoNotUseOutRefParameters.cs]: The code is simple and readable, adding a helper to return Nullable<SyntaxToken> will not improve the situation much and I don't particularly like testing for default(SyntaxToken) too...

**Pull request merged**